File: core/src/com/desitum/library/animation/RotateAnimator.java
package com.desitum.library.animation;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.desitum.library.interpolation.AccelerateDecelerateInterpolator;
import com.desitum.library.interpolation.AccelerateInterpolator;
import com.desitum.library.interpolation.AnticipateInterpolator;
import com.desitum.library.interpolation.BounceInterpolator;
import com.desitum.library.interpolation.DecelerateInterpolator;
import com.desitum.library.interpolation.Interpolation;
import com.desitum.library.interpolation.Interpolator;
import com.desitum.library.interpolation.LinearInterpolator;
import com.desitum.library.interpolation.OvershootInterpolator;

/**
 * Created by kody on 12/21/15.
 * can be used by kody and people in [kody}]
 */
public class RotateAnimator extends Animator {

    private float mAmountToRotate;
    private float mStartRotation;
    private float mEndRotation;
    private float mCurrentRotation;

    private Interpolator mInterpolator;

    public RotateAnimator(float duration, float delay, float amountToRotate, int interpolator) {
        super(duration, delay);
        this.mAmountToRotate = amountToRotate;
        this.mStartRotation = 0;
        this.mEndRotation = amountToRotate;
        setupInterpolator(interpolator);
    }

    public RotateAnimator(Sprite sprite, float duration, float delay, float startRotation, float endRotation, int interpolator) {
        super(sprite, duration, delay);
        this.mStartRotation = startRotation;
        this.mEndRotation = endRotation;
        this.mAmountToRotate = endRotation - startRotation;
        setupInterpolator(interpolator);
    }

    @Override
    public float getCurrentAmount() {
        return 0;
    }

    @Override
    protected void updateAnimation() {
        mCurrentRotation = mStartRotation + mAmountToRotate * mInterpolator.getInterpolation(getTimeInAnimation());

        if (mSprite != null) {
            mSprite.setRotation(mCurrentRotation);
        }
    }

    @Override
    public Animator duplicate() {
        return new RotateAnimator(mSprite, getDuration(), getDelay(), mStartRotation, mEndRotation, Interpolation.getInterpolatorNum(mInterpolator));
    }

    private void setupInterpolator(int interpolator) {
        if (interpolator == Interpolation.ACCELERATE_INTERPOLATOR) {
            this.mInterpolator = AccelerateInterpolator.$();
        } else if (interpolator == Interpolation.DECELERATE_INTERPOLATOR) {
            this.mInterpolator = DecelerateInterpolator.$();
        } else if (interpolator == Interpolation.ANTICIPATE_INTERPOLATOR) {
            this.mInterpolator = AnticipateInterpolator.$();
        } else if (interpolator == Interpolation.OVERSHOOT_INTERPOLATOR) {
            this.mInterpolator = OvershootInterpolator.$();
        } else if (interpolator == Interpolation.ACCELERATE_DECELERATE_INTERPOLATOR) {
            this.mInterpolator = AccelerateDecelerateInterpolator.$();
        } else if (interpolator == Interpolation.BOUNCE_INTERPOLATOR) {
            this.mInterpolator = BounceInterpolator.$();
        } else if (interpolator == Interpolation.LINEAR_INTERPOLATOR) {
            this.mInterpolator = LinearInterpolator.$();
        }
    }
}

File: core/src/com/desitum/library/animation/RotateAnimator.kt
package com.desitum.library.animation

import com.badlogic.gdx.graphics.g2d.Sprite
import com.desitum.library.interpolation.AccelerateDecelerateInterpolator
import com.desitum.library.interpolation.AccelerateInterpolator
import com.desitum.library.interpolation.AnticipateInterpolator
import com.desitum.library.interpolation.BounceInterpolator
import com.desitum.library.interpolation.DecelerateInterpolator
import com.desitum.library.interpolation.Interpolation
import com.desitum.library.interpolation.Interpolator
import com.desitum.library.interpolation.LinearInterpolator
import com.desitum.library.interpolation.OvershootInterpolator

/**
 * Created by kody on 12/21/15.
 * can be used by kody and people in [kody}]
 */
class RotateAnimator : Animator {

    private var amountToRotate: Float = 0f
    private var startRotation: Float = 0f
    private var endRotation: Float = 0f
    var currentRotation: Float = 0f
        private set

    private var interpolator: Interpolator? = null

    constructor(duration: Float, delay: Float, amountToRotate: Float, interpolator: Int) : this(null, duration, delay, 0f, amountToRotate, interpolator)

    constructor(sprite: Sprite?, duration: Float, delay: Float, startRotation: Float, endRotation: Float, interpolator: Int) : super(sprite, duration, delay) {
        this.startRotation = startRotation
        this.endRotation = endRotation
        this.amountToRotate = endRotation - startRotation
        setupInterpolator(interpolator)
    }

    override fun updateAnimation() {
        currentRotation = startRotation + amountToRotate * interpolator!!.getInterpolation(timeInAnimation)

        sprite?.rotation = currentRotation
    }

    override fun duplicate(): Animator {
        return RotateAnimator(sprite, duration, delay, startRotation, endRotation, Interpolation.getInterpolatorNum(interpolator))
    }

    private fun setupInterpolator(interpolator: Int) {
        when (interpolator) {
            Interpolation.ACCELERATE_DECELERATE_INTERPOLATOR -> this.interpolator = AccelerateDecelerateInterpolator.`$`()
            Interpolation.ACCELERATE_INTERPOLATOR -> this.interpolator = AccelerateInterpolator.`$`()
            Interpolation.DECELERATE_INTERPOLATOR -> this.interpolator = DecelerateInterpolator.`$`()
            Interpolation.ANTICIPATE_INTERPOLATOR -> this.interpolator = AnticipateInterpolator.`$`()
            Interpolation.OVERSHOOT_INTERPOLATOR -> this.interpolator = OvershootInterpolator.`$`()
            Interpolation.BOUNCE_INTERPOLATOR -> this.interpolator = BounceInterpolator.`$`()
            Interpolation.LINEAR_INTERPOLATOR -> this.interpolator = LinearInterpolator.`$`()
        }
    }
}